internal/pchan/pchan.go

```diff
@@ -66,26 +68,38 @@ func (ch *PChan[T]) Push(val T, withPrio ...int) chan struct{} {
 
 	go func() { ch.ready <- struct{}{} }()
 
-	return done
-}
-
-// Len returns the number of items queued.
-func (ch *PChan[T]) Len() int {
-	ch.lock.Lock()
-	defer ch.lock.Unlock()
+	ch.wg.Add(1)
 
-	return len(ch.items)
+	return done
 }
 
 // Pop blocks until an item is available, then returns that item.
 // If the channel is already closed, the call returns immediately and the bool value is false.
 func (ch *PChan[T]) Pop() (T, bool) {
 	select {
 	case <-ch.ready:
+		// ...
+
 	case <-ch.done:
+		// ...
+	}
+
+	ch.lock.Lock()
+	defer ch.lock.Unlock()
+
+	if len(ch.items) == 0 {
+		return *new(T), false
```

Comment:
Any particular reason why we can't return nil here?

Reply:
Here we are returning a value of type T, which may or may not be a pointer (it's generic, we don't know what it could be). `nil` only works for pointer-like things (pointers, slices, maps, channels). So we return the default-initialized value of type T by creating a new object of type T on the heap and dereferencing it.

Reply:
But yes, this is bad, shouldn't allocate on the heap just to copy back to the stack... will change to stack-allocated variable.

Reply:
Done, better now